File: contenido/Teoria de Grafos/Kruskal/kruskal.cpp
#include <bits/stdc++.h> 
#define input freopen("in.txt", "r", stdin)
#define output freopen("out.txt", "w", stdout)
// Variables
#define MAX_N 10000    // Maximo numero de vertices o nodos

using namespace std; 

// parent y rango sirven para el union find 
int parent[MAX_N];
int rango[MAX_N];

int n, m; // n numero de nodos y m numero de aristas 

void init() {
    for(int i=0;  i<= MAX_N; i++) {
        parent[i] = i;
        rango[i] = 0;
    }
}

int find(int x) {
    if(x == parent[x]) {
        return x;
    }
    else {
        parent[x] = find(parent[x]);
        return parent[x];
    }
}

bool sameComponent(int nodeA,int nodeB) {
    return find(nodeA) == find(nodeB);
}

void unionRango(int x,int y) {
    int xRaiz = find(x);
    int yRaiz = find(y);
    if(rango[xRaiz] > rango[yRaiz]) {
        parent[yRaiz] = xRaiz;
    } else {
        parent[xRaiz] = yRaiz;
        if(rango[xRaiz] == rango[yRaiz]) {
            rango[yRaiz]++;
        }
    }
}

struct Arista{ 
    int origen;
    int destino; 
    double peso; 
    Arista(){}
    // sobrecarga del operador < para ordenar las aristas
    // sort 
    bool operator<(const Arista &a) const {
        if(peso == a.peso) {
            return origen < a.origen;
        } else {
            return peso < a.peso;
        }
    }
}aristas[MAX_N]; 

Arista MST[MAX_N]; // n-1 aristas 

int numAristasArbol;
double kruskal(int nroNodos, int nroAristas) {
    
    int origen, destino;
    double peso;

    double total = 0; // total del arbol de expansión minimo 

    numAristasArbol = 0; // contar el numero de aristas 

    init(); // Iniciar el union Find 
    // sort(array, array+n)
    // array = [a,b,c,d,e]
    // n = 5
    // sort(array, array+5)
    sort(aristas,aristas + nroAristas); // ordena las aristas
    // sor utiliza el operador < 

    for (int i = 0; i < nroAristas; i++)
    {
        origen = aristas[i].origen;
        destino = aristas[i].destino;
        peso = aristas[i].peso;

        if(!sameComponent(origen,destino)) { // estos 2 dos nodos forman un ciclo 
            total += peso; 
            unionRango(origen,destino);  // unimos los nodos
            MST[numAristasArbol] = aristas[i]; // Guardamos la arista agregada
            numAristasArbol++; // incrementados la posicion para la futura arista
            
        }
    }
    return total;
}

int main(){
    input;
    while(scanf("%d %d",&n,&m) != EOF) {
        for(int i=0; i<m; i++) {
            scanf("%d %d %lf",&aristas[i].origen,&aristas[i].destino,&aristas[i].peso);
        }
        printf("%.2lf\n",kruskal(n,m));
        // Imprimo el arbol de expasion minimo
        for(int i = 0 ;i <numAristasArbol;i++) {
            cout<<MST[i].origen<<" "<<MST[i].destino<<" "<<MST[i].peso<<endl;
        }
    }

    return 0;
}

File: contenido/Teoria de Grafos/Kruskal/Kruskal.java
package teoriaDeGrafos;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Kruskal {
	static int MAX_N = 100010;
	static int[] nodos = new int[10000];
	static int[] padres = new int[10000];
	static int[] rango = new int[10000];
	static int numeroNodos, numeroAristas;

	static void init() {
		for (int i = 0; i <= numeroNodos; i++) {
			padres[i] = i;
			rango[i] = 0;
		}
	}

	static int find(int x) {
		if (x == padres[x]) {
			return x;
		} else {
			padres[x] = find(padres[x]);
			return padres[x];
		}
	}

	static boolean mismoComponente(int nodoA, int nodoB) {
		return find(nodoA) == find(nodoB);
	}

	static void unionRango(int x, int y) {
		int xRaiz = find(x);
		int yRaiz = find(y);
		if (rango[xRaiz] > rango[yRaiz]) {
			padres[yRaiz] = xRaiz;
		} else {
			padres[xRaiz] = yRaiz;
			if (rango[xRaiz] == rango[yRaiz]) {
				rango[yRaiz]++;
			}
		}
	}

	static Arista[] aristas = new Arista[numeroAristas];
	static Arista[] MST = new Arista[MAX_N];

	static void kruskal() {
		int origen, destino, peso;
		int total = 0;
		int numAristasArbol = 0;
		init();
		Arrays.sort(aristas);
		for (int i = 0; i < numeroAristas; i++) {
			origen = aristas[i].origen;
			destino = aristas[i].destino;
			peso = aristas[i].peso;

			if (!mismoComponente(origen, destino)) {
				total += peso;
				MST[numAristasArbol++] = aristas[i];
				unionRango(origen, destino);
			}
		}

		if (numAristasArbol != numeroNodos - 1) {
			System.out.println("no existe un arbol de expansion minima");
		} else {
			for (int i = 0; i < numeroNodos - 1; i++) {
				System.out.print(MST[i].origen+" ");
				System.out.print(MST[i].destino+" ");
				System.out.println(MST[i].peso);
			}
			System.out.println(total);
		}
	}

	public static void main(String[] args) throws IOException {
		BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
		numeroNodos=Integer.parseInt(bf.readLine());
		numeroAristas=Integer.parseInt(bf.readLine());
		aristas =new Arista[numeroAristas]; 
		for (int i = 0; i < numeroAristas; i++) {
			String[] linea= bf.readLine().split(" ");
			int inicio=Integer.parseInt(linea[0]);
			int fin=Integer.parseInt(linea[1]);
			int peso=Integer.parseInt(linea[2]);
			Arista nuevaArista=new Arista(inicio, fin, peso);
			aristas[i]=nuevaArista;
		}
		kruskal();
	}
}

class Arista implements Comparable<Arista> {
	int origen;
	int destino;
	int peso;

	public Arista(int origen, int destino, int peso) {
		this.origen = origen;
		this.destino = destino;
		this.peso = peso;
	}

	@Override
	public int compareTo(Arista o) {
		// TODO Auto-generated method stub
		return this.peso - o.peso;
	}

}
